### PR TITLE
Implement retry logic in ohsnap helper methods using wait-for library

### PR DESCRIPTION
**Pull request by rplevka**
This commit should cover us from the intermittent, short outages of `ohsnap` service.
It leverages the `wait-for` library that is used in a way that if the outage persists after the method timing out, an original Error is thrown.
In order to be able to use it this way, I had to register a [requests hook](https://docs.python-requests.org/en/latest/user/advanced/#event-hooks), that automatically runs `raise_for_status` on every response and logs the unsuccessful attempts.

Test:
I spawned my own instnace of ohsnap and simulated outage by scaling it down to 0 pods, and then back to 1 pod after a while:
```
2023-02-28 13:18:56 - urllib3.connectionpool - DEBUG - https://vault.corp.redhat.com:8200 "GET /v1/apps/data/mna-satqe-hashicorp/robottelo HTTP/1.1" 200 None
2023-02-28 13:18:58 - robottelo - WARNING - The snap version was not provided. Snap number will not be used in the URL.
2023-02-28 13:18:58 - urllib3.connectionpool - DEBUG - Starting new HTTP connection (1): ohsnap-instance.foo:80
2023-02-28 13:18:58 - urllib3.connectionpool - DEBUG - http://ohsnap-instance.foo:80 "GET /api/releases/6.13.0/el8/capsule/repositories HTTP/1.1" 503 None
2023-02-28 13:18:58 - robottelo - WARNING - Request to: http://ohsnap-instance.foo/api/releases/6.13.0/el8/capsule/repositories got response: 503
2023-02-28 13:19:01 - robottelo - WARNING - The snap version was not provided. Snap number will not be used in the URL.
2023-02-28 13:19:01 - urllib3.connectionpool - DEBUG - Starting new HTTP connection (1): ohsnap-instance.foo:80
2023-02-28 13:19:01 - urllib3.connectionpool - DEBUG - http://ohsnap-instance.foo:80 "GET /api/releases/6.13.0/el8/capsule/repositories HTTP/1.1" 503 None
2023-02-28 13:19:01 - robottelo - WARNING - Request to: http://ohsnap-instance.foo/api/releases/6.13.0/el8/capsule/repositories got response: 503
2023-02-28 13:19:04 - robottelo - WARNING - The snap version was not provided. Snap number will not be used in the URL.
2023-02-28 13:19:04 - urllib3.connectionpool - DEBUG - Starting new HTTP connection (1): ohsnap-instance.foo:80
2023-02-28 13:19:04 - urllib3.connectionpool - DEBUG - http://ohsnap-instance.foo:80 "GET /api/releases/6.13.0/el8/capsule/repositories HTTP/1.1" 503 None
2023-02-28 13:19:04 - robottelo - WARNING - Request to: http://ohsnap-instance.foo/api/releases/6.13.0/el8/capsule/repositories got response: 503
2023-02-28 13:19:07 - robottelo - WARNING - The snap version was not provided. Snap number will not be used in the URL.
2023-02-28 13:19:07 - urllib3.connectionpool - DEBUG - Starting new HTTP connection (1): ohsnap-instance.foo:80
2023-02-28 13:19:08 - urllib3.connectionpool - DEBUG - http://ohsnap-instance.foo:80 "GET /api/releases/6.13.0/el8/capsule/repositories HTTP/1.1" 503 None
2023-02-28 13:19:08 - robottelo - WARNING - Request to: http://ohsnap-instance.foo/api/releases/6.13.0/el8/capsule/repositories got response: 503
2023-02-28 13:19:11 - robottelo - WARNING - The snap version was not provided. Snap number will not be used in the URL.
2023-02-28 13:19:11 - urllib3.connectionpool - DEBUG - Starting new HTTP connection (1): ohsnap-instance.foo:80
2023-02-28 13:19:11 - urllib3.connectionpool - DEBUG - http://ohsnap-instance.foo:80 "GET /api/releases/6.13.0/el8/capsule/repositories HTTP/1.1" 503 None
2023-02-28 13:19:11 - robottelo - WARNING - Request to: http://ohsnap-instance.foo/api/releases/6.13.0/el8/capsule/repositories got response: 503
...
...
...
# scaled back to 1 pod
2023-02-28 13:19:54 - urllib3.connectionpool - DEBUG - Starting new HTTP connection (1): ohsnap-instance.foo:80
2023-02-28 13:20:17 - urllib3.connectionpool - DEBUG - http://ohsnap-instance.foo:80 "GET /api/releases/6.13.0/el8/satellite/repositories HTTP/1.1" 200 None
2023-02-28 13:20:17 - urllib3.connectionpool - DEBUG - Starting new HTTP connection (1): dogfood.server.foo:80
2023-02-28 13:20:17 - urllib3.connectionpool - DEBUG - http://dogfood.server.foo:80 "GET /pulp/content/Satellite_Engineering/QA/Satellite_6_13_RHEL8/custom/Satellite_6_13_Composes/Satellite_6_13_RHEL8/ HTTP/1.1" 200 154
```

### CODE REVIEW

File: robottelo/utils/ohsnap.py [hunk 1]
```diff
@@ -45,20 +53,26 @@ def dogfood_repository(
...
-    if repo_check and not requests.get(repository['baseurl']).ok:
+    dogfood_req = requests.get(repository['baseurl'])
+    if repo_check and not dogfood_req.ok:
         logger.warn(
-            f'Repository was not found on the URL: {repository["baseurl"]} ; Arguments used: '
-            f'repo={repo}, product={product}, release={release}, os_release={os_release}, '
-            f'snap={snap}'
+            f'Unable to locate the repo at the URL: {repository["baseurl"]} ; '
+            f'HTTP response: {dogfood_req.status_code}; Arguments used: repo={repo}, '
+            f'product={product}, release={release}, os_release={os_release}, snap={snap}'
```
Comment (ogajduse) [inline robottelo/utils/ohsnap.py:76]:
```suggestion
            f'Unable to locate the repo at the URL: {repository["baseurl"]} ; '
            f'HTTP response: {dogfood_req.status_code}; Arguments used: {repo=}, '
            f'{product=}, {release=}, {os_release=}, {snap=}'
```

**Comment by ogajduse** [comment]
trigger: test-robottelo
pytest: tests/foreman/api/test_ping.py

**Review by JacobCallahan — APPROVED** [review]

**Review by JacobCallahan — APPROVED** [review]

File: conf/ohsnap.yaml.template [hunk 2]
```diff
@@ -0,0 +1,7 @@
+OHSNAP:
+  HOST: "replace by ohsnap repo host url"
+  REQUEST_RETRY:
```
Comment (Gauravtalreja1) [inline conf/ohsnap.yaml.template:3]: Could you remove a old `OHSNAP_REPO_HOST` from template and config validators?

File: robottelo/utils/ohsnap.py [hunk 3]
```diff
@@ -2,13 +2,21 @@
...
 from robottelo import constants
 from robottelo.exceptions import InvalidArgumentError
 from robottelo.exceptions import RepositoryDataNotFound
 from robottelo.logging import logger
 
 
+def ohsnap_response_hook(r, *args, **kwargs):
+    """Requests response hook callback function that processes the response"""
+    if not r.ok:
+        logger.warn(f'Request to: {r.request.url} got response: {r.status_code}')
```
Comment (Gauravtalreja1) [inline robottelo/utils/ohsnap.py:16]: warn is a deprecated method, can we use warning?
```suggestion
        logger.warning(f'Request to: {r.request.url} got response: {r.status_code}')
```

File: robottelo/utils/ohsnap.py [hunk 4]
```diff
@@ -41,24 +49,31 @@ def dogfood_repofile_url(ohsnap_repo_host, product, release, os_release, snap=''
...
     try:
         repository = next(r for r in res.json() if r['label'] == repo)
     except StopIteration:
         raise RepositoryDataNotFound(f'Repository "{repo}" is not provided by the given product')
     repository['baseurl'] = repository['baseurl'].replace('$basearch', arch)
     # If repo check is enabled, check that the repository actually exists on the remote server
-    if repo_check and not requests.get(repository['baseurl']).ok:
+    dogfood_req = requests.get(repository['baseurl'])
+    if repo_check and not dogfood_req.ok:
         logger.warn(
```
Comment (Gauravtalreja1) [inline robottelo/utils/ohsnap.py:74]: same here

**Review by Gauravtalreja1 — COMMENTED** [review]
Ack, with a few comments/suggestion

File: robottelo/utils/ohsnap.py (hunk 3 above)
Comment (rplevka) [reply robottelo/utils/ohsnap.py:16]: ah darn, thought it was the other way around :D

File: conf/dynaconf_hooks.py [hunk 5]
```diff
@@ -127,7 +127,7 @@ def get_dogfood_satclient_repos(settings):
 
 def get_ohsnap_repo_url(settings, repo, product=None, release=None, os_release=None, snap=''):
     repourl = dogfood_repository(
-        settings.repos.ohsnap_repo_host,
+        settings.ohsnap,
```
[the remark this answers is not in the document]
Comment (rplevka) [reply conf/dynaconf_hooks.py:130]: done

File: conf/ohsnap.yaml.template (hunk 2 above)
Comment (rplevka) [reply conf/ohsnap.yaml.template:3]: done

File: robottelo/utils/ohsnap.py (hunk 3 above)
Comment (rplevka) [reply robottelo/utils/ohsnap.py]: done

File: robottelo/utils/ohsnap.py (hunk 4 above)
Comment (rplevka) [reply robottelo/utils/ohsnap.py]: done

**Review by ogajduse — APPROVED** [review]
ACK, pending @Gauravtalreja1's comments

File: conf/dynaconf_hooks.py (hunk 5 above)
Comment (Gauravtalreja1) [inline conf/dynaconf_hooks.py:130]:
```suggestion
       settings.ohsnap.host,
```

File: robottelo/host_helpers/contenthost_mixins.py [hunk 6]
```diff
@@ -72,15 +72,13 @@ def _dogfood_helper(self, product, release, snap, repo=None):
...
+        url = dogfood_repofile_url(settings.ohsnap.host, product, release, v_major, snap)
         self.execute(f'curl -o /etc/yum.repos.d/dogfood.repo {url}')
 
     def dogfood_repository(self, repo=None, product=None, release=None, snap=''):
         """Returns a repository definition based on the arguments provided"""
         product, release, snap, v_major, repo = self._dogfood_helper(product, release, snap, repo)
-        return dogfood_repository(
-            settings.repos.ohsnap_repo_host, repo, product, release, v_major, snap, self.arch
-        )
+        return dogfood_repository(settings.ohsnap, repo, product, release, v_major, snap, self.arch)
```
Comment (Gauravtalreja1) [inline robottelo/host_helpers/contenthost_mixins.py:81]:
```suggestion
        return dogfood_repository(settings.ohsnap.host, repo, product, release, v_major, snap, self.arch)
```

File: tests/foreman/api/test_remoteexecution.py [hunk 7]
```diff
@@ -100,7 +100,7 @@ def test_negative_time_to_pickup(
     :parametrized: yes
     """
     client_repo = ohsnap.dogfood_repository(
-        settings.repos.ohsnap_repo_host,
+        settings.ohsnap,
```
Comment (Gauravtalreja1) [inline tests/foreman/api/test_remoteexecution.py:103]:
```suggestion
        settings.ohsnap.host,
```

**Review by Gauravtalreja1 — COMMENTED** [review]
`s/settings.repos.ohsnap_repo_host/settings.ohsnap.host` :)

File: conf/dynaconf_hooks.py (hunk 5 above)
Comment (rplevka) [reply conf/dynaconf_hooks.py:130]: @Gauravtalreja1 , no, this one is correct. The method now accepts the `ohsnap` dictionary as it uses the rest of the settings further.
Same applies to your other suggestions

File: robottelo/utils/ohsnap.py [hunk 8]
```diff
@@ -2,13 +2,21 @@
...
+from wait_for import wait_for
 
 from robottelo import constants
 from robottelo.exceptions import InvalidArgumentError
 from robottelo.exceptions import RepositoryDataNotFound
 from robottelo.logging import logger
 
 
+def ohsnap_response_hook(r, *args, **kwargs):
+    """Requests response hook callback function that processes the response"""
```
Comment (jyejare) [inline robottelo/utils/ohsnap.py:14]: Expand on docstrings for parameters atleast for `r`.

**Review by jyejare — APPROVED** [review]

**Review by jyejare — CHANGES_REQUESTED** [review]
PRT is failing with :
`settings object not found ohsnap attr`, please check and resolve it.

File: robottelo/utils/ohsnap.py (hunk 8 above)
Comment (rplevka) [reply robottelo/utils/ohsnap.py]: done

**Review by jyejare — APPROVED** [review]

**Pull request merged**